### PR TITLE
Create did

### PR DESCRIPTION
**Pull request by Balurat**
This PR addresses issue #1 and #4 (in part) but is meant to discuss some questions I have. The implementation is **not yet finished**.

In order to test the code on the Jungle Testnet, I have created an account, tapped the faucet for some EOS, and staked it for resources. The key material is saved locally to a JSON-file, which I added to the .gitignore. That way the private key never ends up on the repo and others could use custom accounts for testing.

**The test runs fine** and the account gets created 👍 
So far so good.
\
\
But the resulting `DIDDocument` is little more than a placeholder. I was wondering if we should _refactor_ the `createDIDDocument` function from the resolver so both packages can use it and we can maintain the logic in one place.
\
\
_I'm uncertain about validation and error handling:_
Which (if any) inputs should I validate? For now, I just check if the account names and chain id are to spec.

Is the absence of errors enough to know that the transaction went through as planned? Should I check that the `error_code` field is `null` or something like that?
\
\
It seems to me, that we need some **additional parameter** for the function.
Because of that, I changed the function signature:
```js
export default async function create(
  chain: string,
  creator: string,
  name: string,
  owner: Authority,
  active: Authority,
  signatureProvider: SignatureProvider,
  options?: ConfigOptions
): Promise<DIDDocument>
```
The `chain` contains either the chain name or id. The `signatureProvider` holds the private key of the creator.

There are also a few fields of the transaction that a user might want to override. I moved those into an `additionalOptions` object for clarity. 
As far as I understand you have to buy RAM and delegate some CPU and NET for the account creation to work.
```js
const additionalOptions = {
    receiverAccount: 'eosio',  // I don't know if chains can customize this name
    authorization: [
      {
        actor: creator,
        permission: 'active',
      },
    ],
    buyrambytes: {
      bytes: 8192,
    },
    delegatebw: {
      stakeNetQuantity: '1.0000 EOS',
      stakeCpuQuantity: '1.0000 EOS',
      transfer: false,
    },
    config: {
      blocksBehind: 3,
      expireSeconds: 30,
    },
  };
```
Do you think the whole transaction should be somewhat reconfigurable?
Maybe instead of using the `options` parameter solely for custom chain registry entries, we could expand it? 
So the type could look something like this:
```js
export declare type ConfigOptions = {
    registry?: ChainRegistry,
    authorization?: [
        {
          actor: string,
          permission: string,
        },
    ], 
    create?: {
        buyrambytes?: {
          bytes: number,
        },
        delegatebw?: {
          stakeNetQuantity?: string,
          stakeCpuQuantity?: string,
          transfer?: boolean,
        },
    },
    config?: {
        blocksBehind?: number,
        expireSeconds?: number,
    },
    [x: string]: any
}
export declare type ChainRegistry = {
    [x: string]: {
        chainId: string,
        service: [{
            id: string,
            type: string | [string],
            serviceEndpoint: string
        }]
    }
}
```
If we choose to expand the options object, we could also move some or all of the other parameters in there. I'm unclear if this is useful or unnecessary.
\
\
A _small change_ was made to the **type Authority** because NodeEOS threw an error whenever keys and/or waits were not set. So I made them mandatory and allowed for an empty array.
\
\
Finally, I ran `prettier` over the whole project 🥇 
(In reference to PR [#9](https://github.com/Gimly-Blockchain/eosio-did-resolver/pull/9) from the resolver repo)
\
\
**So in summary:**

- [x] Decide how to handle options
- [x] Decide whether to validate inputs before sending the transaction
- [x] Decide whether to validate the transaction result
- [ ] Implement a proper DID document result / refactor `createDIDDocument()`

Any other suggestions or issues?

### CODE REVIEW

**Comment by theblockstalk** [comment]
> I was wondering if we should refactor the createDIDDocument function from the resolver so both packages can use it and we can maintain the logic in one place.

yes we should do this to avoid duplicate code in both repositories

_______________
> Which (if any) inputs should I validate?

I think none is ok, as the tx will fail if they incorrect.

_______________
> Is the absence of errors enough to know that the transaction went through as planned? Should I check that the error_code field is null or something like that?

I suggest you copy what is done in the section "Error handling" in the https://www.npmjs.com/package/eosjs documentation
Leave it at that, if they want more control they can implement it themself or we can handle it later. for now I think this is adequate

_______________
I think it is a good idea to expand the options parameter. This is what I think should be customizeable:
```js
export declare type ConfigOptions = {
    registry?: ChainRegistry,
   receiverAccount?: sting,
   creatorPermission?: string,
   buyrambytes?: number,
   stakeNetQuantity?: string,
   stakeCpuQuantity?: string
}
```

_______________
Could you make some suggestions of what we need to change or expose in the resolver to reduce duplication?

**Comment by Balurat** [comment]
Made the changes we talked about. 

In order to import the chain name registry, I made a [PR](https://github.com/Gimly-Blockchain/eosio-did-resolver/pull/27) to the did-resolver repo to add the necessary exports.
The regex patterns were also moved to the did-resolver to avoid duplication.

The resulting DIDDocument of the create function now is generated through the resolve function.

**Comment by theblockstalk** [comment]
It's looking very good. A few questions:
1. Does DID create return a DID Document result?
2. Does this already implement DID update? it looks like it
3. It looks like this depends on the global `fetch` object. This wont work for nodejs so we'll need to provide a way to give the `require('node-fetch')` instead same as what we did in the resolver

Regardless of the above answers, I'd be happy to resolve this now and then figure out the above as necessary in the next PR

File: src/create.ts [hunk 1]
```diff
@@ -1,12 +1,171 @@
...
+    ...options?.registry,
+  };
+
+  const chainData = getChainData(chainRegistry, chain);
+
+  let fetchAttempts = 0;
+  for (const service of chainData.service) {
+    fetchAttempts++;
+
+    const rpc = new JsonRpc(service.serviceEndpoint, { fetch });
```
Comment (theblockstalk) [inline src/create.ts:63]: this is using the `node-fetch` package which isn't necessary for browsers. I suggest we make this part of the ConfigOptions
